In [5]:
import pandas as pd
df=pd.read_csv("/content/movies.csv")
df.head(5)

,MOVIES,YEAR,GENRE,RATING,ONE-LINE,STARS,VOTES,RunTime,Gross
0,Blood Red Sky,(2021),"\nAction, Horror, Thriller",6.1,\nA woman with a mysterious illness is forced ...,\n Director:\nPeter Thorwarth\n| \n Star...,"21,062",121.0,NaN
1,Masters of the Universe: Revelation,(2021– ),"\nAnimation, Action, Adventure",5.0,\nThe war for Eternia begins again in what may...,"\n \n Stars:\nChris Wood, \nSara...","17,870",25.0,NaN
2,The Walking Dead,(2010–2022),"\nDrama, Horror, Thriller",8.2,\nSheriff Deputy Rick Grimes wakes up from a c...,"\n \n Stars:\nAndrew Lincoln, \n...","885,805",44.0,NaN
3,Rick and Morty,(2013– ),"\nAnimation, Adventure, Comedy",9.2,\nAn animated series that follows the exploits...,"\n \n Stars:\nJustin Roiland, \n...","414,849",23.0,NaN
4,Army of Thieves,(2021),"\nAction, Crime, Horror",NaN,"\nA prequel, set before the events of Army of ...",\n Director:\nMatthias Schweighöfer\n| \n ...,NaN,NaN,NaN


In [19]:
df.shape

(9999, 9)

In [6]:
#data cleaning
#1)finding nulls in columns and filling them
df.isnull().sum()

,0
MOVIES,0
YEAR,644
GENRE,80
RATING,1820
ONE-LINE,0
STARS,0
VOTES,1820
RunTime,2958
Gross,9539


if categorical then fill the nulls by mode and if numerical then fill nulls by mean

In [32]:
import pandas as pd

df['GENRE']=df['GENRE'].fillna(df['GENRE'].mode()[0])
df['YEAR']=df['YEAR'].fillna(df['YEAR'].mode()[0])
df['RATING']=df['RATING'].fillna(df['RATING'].mean())

# Check if 'VOTES' is still a string before applying string operations
if pd.api.types.is_string_dtype(df['VOTES']):
    df['VOTES'] = df['VOTES'].str.replace(',', '', regex=False).astype(float)
df['VOTES']=df['VOTES'].fillna(df['VOTES'].mean())
df['RunTime']=df['RunTime'].fillna(df['RunTime'].mean())




In [31]:
# The 'Gross' column was already dropped in a previous step.
df.isnull().sum()

,0
MOVIES,0
YEAR,0
GENRE,0
RATING,0
ONE-LINE,0
STARS,0
VOTES,0
RunTime,0


In [37]:
#removing duplicates
df.drop_duplicates(inplace=True)

In [35]:
df.isnull().sum()

,0
MOVIES,0
YEAR,0
GENRE,0
RATING,0
ONE-LINE,0
STARS,0
VOTES,0
RunTime,0


In [36]:
df.duplicated().sum()

np.int64(0)

In [38]:
df.head(5)

,MOVIES,YEAR,GENRE,RATING,ONE-LINE,STARS,VOTES,RunTime
0,Blood Red Sky,(2021),"\nAction, Horror, Thriller",6.100000,\nA woman with a mysterious illness is forced ...,\n Director:\nPeter Thorwarth\n| \n Star...,21062.000000,121.000000
1,Masters of the Universe: Revelation,(2021– ),"\nAnimation, Action, Adventure",5.000000,\nThe war for Eternia begins again in what may...,"\n \n Stars:\nChris Wood, \nSara...",17870.000000,25.000000
2,The Walking Dead,(2010–2022),"\nDrama, Horror, Thriller",8.200000,\nSheriff Deputy Rick Grimes wakes up from a c...,"\n \n Stars:\nAndrew Lincoln, \n...",885805.000000,44.000000
3,Rick and Morty,(2013– ),"\nAnimation, Adventure, Comedy",9.200000,\nAn animated series that follows the exploits...,"\n \n Stars:\nJustin Roiland, \n...",414849.000000,23.000000
4,Army of Thieves,(2021),"\nAction, Crime, Horror",6.921176,"\nA prequel, set before the events of Army of ...",\n Director:\nMatthias Schweighöfer\n| \n ...,15124.062722,68.688539


In [39]:
#fixing inconsistent data
df['YEAR']

,YEAR
0,(2021)
1,(2021– )
2,(2010–2022)
3,(2013– )
4,(2021)
...,...
9993,(2022– )
9995,(2021– )
9996,(2022– )
9997,(2021– )


Year column contained String values , in order to extract Year and convert it into formatted year colum n , we needed to convert the type String to float



In [53]:
#4-digit year
df['YEAR'] = df['YEAR'].astype(str).str.extract(r'(\d{4})')

# string to numeric
df['YEAR'] = pd.to_numeric(df['YEAR'], errors='coerce')

# Convert to integer
df['YEAR'] = df['YEAR'].astype(int)

print(df['YEAR'].value_counts())


YEAR
2020    2311
2019    1334
2018    1074
2021     987
2017     835
        ... 
1957       1
1972       1
1944       1
1933       1
1938       1
Name: count, Length: 78, dtype: int64


In [54]:
df['GENRE'] = df['GENRE'].str.strip() #trims leading and trailing blank spaces
df['GENRE'] = df['GENRE'].str.replace('\n', '', regex=False) #removes newline characters to get uniqye value counts

print(df['GENRE'].value_counts())

GENRE
Comedy                          915
Animation, Action, Adventure    663
Drama                           546
Documentary                     495
Crime, Drama, Mystery           328
                               ... 
Short, Reality-TV                 1
Short, Action                     1
Family, Music                     1
News, Reality-TV                  1
Music, Musical                    1
Name: count, Length: 510, dtype: int64


In [55]:
df['Director'] = df['STARS'].str.extract(r'Director:\n\s*(.*?)(?:\n\| |\n\s*Stars:|$)')
df['Stars'] = df['STARS'].str.extract(r'Stars:\n\s*(.*)')

df['Director'] = df['Director'].str.strip().str.replace('\n', '', regex=False)
df['Stars'] = df['Stars'].str.strip().str.replace('\n', '', regex=False)

print("DataFrame with extracted 'Director' and 'Stars' columns:")
df.head()

DataFrame with extracted 'Director' and 'Stars' columns:


,MOVIES,YEAR,GENRE,RATING,ONE-LINE,STARS,VOTES,RunTime,Director,Stars
0,Blood Red Sky,2021,"Action, Horror, Thriller",6.100000,\nA woman with a mysterious illness is forced ...,\n Director:\nPeter Thorwarth\n| \n Star...,21062.000000,121.000000,Peter Thorwarth,"Peri Baumeister,"
1,Masters of the Universe: Revelation,2021,"Animation, Action, Adventure",5.000000,\nThe war for Eternia begins again in what may...,"\n \n Stars:\nChris Wood, \nSara...",17870.000000,25.000000,NaN,"Chris Wood,"
2,The Walking Dead,2010,"Drama, Horror, Thriller",8.200000,\nSheriff Deputy Rick Grimes wakes up from a c...,"\n \n Stars:\nAndrew Lincoln, \n...",37471.531804,44.000000,NaN,"Andrew Lincoln,"
3,Rick and Morty,2013,"Animation, Adventure, Comedy",9.200000,\nAn animated series that follows the exploits...,"\n \n Stars:\nJustin Roiland, \n...",37471.531804,23.000000,NaN,"Justin Roiland,"
4,Army of Thieves,2021,"Action, Crime, Horror",6.921176,"\nA prequel, set before the events of Army of ...",\n Director:\nMatthias Schweighöfer\n| \n ...,15124.062722,68.688539,Matthias Schweighöfer,"Matthias Schweighöfer,"


In [56]:
#IQR used for finding outliers in rating and vites
for col in ['RATING', 'VOTES', 'RunTime']:
    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)
#this can be also done using df.describe()
    IQR = Q3 - Q1

    upper_bound = Q3 + 1.5 * IQR
    lower_bound = Q1 - 1.5 * IQR

    df[col] = df[col].clip(lower=lower_bound, upper=upper_bound)
    print(f"Outliers in '{col}' capped. New descriptive statistics:")
    print(df[col].describe())
    print("\n")

Outliers in 'RATING' capped. New descriptive statistics:
count    9568.000000
mean        6.948615
std         1.045531
min         4.600000
25%         6.400000
50%         6.921176
75%         7.600000
max         9.400000
Name: RATING, dtype: float64


Outliers in 'VOTES' capped. New descriptive statistics:
count     9568.000000
mean      6900.768592
std      10403.216185
min          5.000000
25%        225.750000
50%       1323.000000
75%      15124.062722
max      37471.531804
Name: VOTES, dtype: float64


Outliers in 'RunTime' capped. New descriptive statistics:
count    9568.000000
mean       67.018673
std        30.390775
min         1.000000
25%        45.000000
50%        68.688539
75%        87.000000
max       150.000000
Name: RunTime, dtype: float64




In [44]:
df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
Index: 9568 entries, 0 to 9998
Data columns (total 10 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   MOVIES    9568 non-null   object 
 1   YEAR      9568 non-null   int64  
 2   GENRE     9568 non-null   object 
 3   RATING    9568 non-null   float64
 4   ONE-LINE  9568 non-null   object 
 5   STARS     9568 non-null   object 
 6   VOTES     9568 non-null   float64
 7   RunTime   9568 non-null   float64
 8   Director  5214 non-null   object 
 9   Stars     8365 non-null   object 
dtypes: float64(3), int64(1), object(6)
memory usage: 822.2+ KB


,MOVIES,YEAR,GENRE,RATING,ONE-LINE,STARS,VOTES,RunTime,Director,Stars
0,Blood Red Sky,2021,"Action, Horror, Thriller",6.100000,\nA woman with a mysterious illness is forced ...,\n Director:\nPeter Thorwarth\n| \n Star...,21062.000000,121.000000,Peter Thorwarth,"Peri Baumeister,"
1,Masters of the Universe: Revelation,2021,"Animation, Action, Adventure",5.000000,\nThe war for Eternia begins again in what may...,"\n \n Stars:\nChris Wood, \nSara...",17870.000000,25.000000,NaN,"Chris Wood,"
2,The Walking Dead,2010,"Drama, Horror, Thriller",8.200000,\nSheriff Deputy Rick Grimes wakes up from a c...,"\n \n Stars:\nAndrew Lincoln, \n...",37471.531804,44.000000,NaN,"Andrew Lincoln,"
3,Rick and Morty,2013,"Animation, Adventure, Comedy",9.200000,\nAn animated series that follows the exploits...,"\n \n Stars:\nJustin Roiland, \n...",37471.531804,23.000000,NaN,"Justin Roiland,"
4,Army of Thieves,2021,"Action, Crime, Horror",6.921176,"\nA prequel, set before the events of Army of ...",\n Director:\nMatthias Schweighöfer\n| \n ...,15124.062722,68.688539,Matthias Schweighöfer,"Matthias Schweighöfer,"


In [59]:
df.to_csv('cleaned_movies.csv', index=False)

## Preprocessing Report for Movies Dataset

This report details the preprocessing steps applied to the `df` DataFrame, which was loaded from the 'movies.csv' file. The goal of these steps was to clean the data, handle inconsistencies, and prepare it for further analysis.

### 1. Initial Data Loading
The dataset was initially loaded into a pandas DataFrame from the file `/content/movies.csv`.

### 2. Handling Missing Values
Missing values were identified using `df.isnull().sum()` and addressed as follows:
-   **'GENRE'**: Missing values were filled using the mode (most frequent value) of the column, as it is a categorical feature.
-   **'YEAR'**: Missing values were also filled using the mode of the column, treated as a categorical feature before type conversion.
-   **'RATING'**: Missing values were imputed using the mean of the column, as it is a numerical feature.
-   **'VOTES'**: This column initially contained string values with commas. Commas were first removed, and the column was converted to a float type. Subsequently, any remaining missing values were filled with the mean of the column.
-   **'RunTime'**: Missing values were filled using the mean of the column.
-   **'Gross'**: It was noted that the 'Gross' column had a very high number of missing values (9539 out of 9999 entries) and was effectively removed from consideration or dropped in an earlier implicit step, as it was not present after the initial null filling.

### 3. Removing Duplicates
Duplicate rows within the DataFrame were identified and removed using `df.drop_duplicates(inplace=True)`, ensuring each record is unique.

### 4. Cleaning and Type Conversion for 'YEAR'
The 'YEAR' column contained string values with extra characters (e.g., '(2021)', '(2021– )'). The cleaning process involved:
-   Extracting the 4-digit year using a regular expression `r'(\d{4})'`.
-   Converting the extracted year strings to a numeric type, coercing errors to NaN.
-   Finally, converting the numeric 'YEAR' column to an integer type.

### 5. Cleaning 'GENRE'
The 'GENRE' column was cleaned to remove inconsistencies and prepare it for analysis:
-   Leading and trailing blank spaces were trimmed using `.str.strip()`.
-   Newline characters (\n) were removed using `.str.replace('\n', '', regex=False)`.

### 6. Extraction of 'Director' and 'Stars'
Information for 'Director' and 'Stars' was embedded within the 'STARS' column. Regular expressions were used to extract this information into new, dedicated columns:
-   **'Director'**: Extracted using `r'Director:\n\s*(.*?)(?:\n\| |\n\s*Stars:|$)'`.
-   **'Stars'**: Extracted using `r'Stars:\n\s*(.*)'`.
-   Both new columns ('Director' and 'Stars') had leading/trailing spaces trimmed and newline characters removed for cleanliness.

### 7. Outlier Treatment
Outliers in the numerical columns 'RATING', 'VOTES', and 'RunTime' were treated using the Interquartile Range (IQR) method:
-   For each column, the first quartile (Q1) and third quartile (Q3) were calculated.
-   The IQR was determined as Q3 - Q1.
-   Upper bounds (Q3 + 1.5 * IQR) and lower bounds (Q1 - 1.5 * IQR) were established.
-   Values falling outside these bounds were capped to the respective upper or lower bound, effectively minimizing the impact of extreme outliers while retaining the data points.


## Display Final DataFrame Info

### Subtask:
Display the .info() and .head() of the final, cleaned DataFrame to show its structure and a sample of the processed data.


**Reasoning**:
The subtask requires displaying the DataFrame's info and head to verify the cleaning steps. The `df.info()` method provides a concise summary of the DataFrame, including data types and non-null values, which is essential after data cleaning. The `df.head()` method will show the first few rows of the processed data, allowing for a quick visual inspection.



In [60]:
df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
Index: 9568 entries, 0 to 9998
Data columns (total 10 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   MOVIES    9568 non-null   object 
 1   YEAR      9568 non-null   int64  
 2   GENRE     9568 non-null   object 
 3   RATING    9568 non-null   float64
 4   ONE-LINE  9568 non-null   object 
 5   STARS     9568 non-null   object 
 6   VOTES     9568 non-null   float64
 7   RunTime   9568 non-null   float64
 8   Director  5214 non-null   object 
 9   Stars     8365 non-null   object 
dtypes: float64(3), int64(1), object(6)
memory usage: 822.2+ KB


,MOVIES,YEAR,GENRE,RATING,ONE-LINE,STARS,VOTES,RunTime,Director,Stars
0,Blood Red Sky,2021,"Action, Horror, Thriller",6.100000,\nA woman with a mysterious illness is forced ...,\n Director:\nPeter Thorwarth\n| \n Star...,21062.000000,121.000000,Peter Thorwarth,"Peri Baumeister,"
1,Masters of the Universe: Revelation,2021,"Animation, Action, Adventure",5.000000,\nThe war for Eternia begins again in what may...,"\n \n Stars:\nChris Wood, \nSara...",17870.000000,25.000000,NaN,"Chris Wood,"
2,The Walking Dead,2010,"Drama, Horror, Thriller",8.200000,\nSheriff Deputy Rick Grimes wakes up from a c...,"\n \n Stars:\nAndrew Lincoln, \n...",37471.531804,44.000000,NaN,"Andrew Lincoln,"
3,Rick and Morty,2013,"Animation, Adventure, Comedy",9.200000,\nAn animated series that follows the exploits...,"\n \n Stars:\nJustin Roiland, \n...",37471.531804,23.000000,NaN,"Justin Roiland,"
4,Army of Thieves,2021,"Action, Crime, Horror",6.921176,"\nA prequel, set before the events of Army of ...",\n Director:\nMatthias Schweighöfer\n| \n ...,15124.062722,68.688539,Matthias Schweighöfer,"Matthias Schweighöfer,"
